iocage/lib/Jail.py

```diff
@@ -0,0 +1,500 @@
+import Storage
+import Releases
+import Release
+import helpers
+
+import ZFSBasejailStorage
+import ZFSShareStorage
+import NullFSBasejailStorage
+import StandaloneJailStorage
+
+import subprocess
+import uuid
+
+
+class Jail:
+
+    def __init__(self, data={}, zfs=None, host=None, logger=None, new=False):
+
+        helpers.init_logger(self, logger)
+        helpers.init_zfs(self, zfs)
+        helpers.init_host(self, host)
+
+        if isinstance(data, str):
+            data = {"uuid": self._resolve_uuid(data)}
+
+        self.config = JailConfig.JailConfig(
+            data=data, jail=self, logger=self.logger)
+
+        self.networks = []
+
+        self.storage = Storage.Storage(
+            auto_create=True, safe_mode=False,
+            jail=self, logger=self.logger, zfs=self.zfs)
+
+        self.jail_state = None
+
+        if new is False:
+            self.config.read()
+        # self.update_jail_state()
+
+    @property
+    def zfs_pool_name(self):
+        return self.host.datasets.root.name.split("/", maxsplit=1)[0]
+
+    def start(self):
+        self.require_jail_existing()
+        self.require_jail_stopped()
+
+        release = self.release
+
+        NullFSBasejailStorage.NullFSBasejailStorage.umount_nullfs(
+            self.storage)
+
+        storage_backend = None
+
+        if self.config.basejail_type == "zfs":
+            storage_backend = ZFSBasejailStorage.ZFSBasejailStorage
+
+        if self.config.basejail_type == "nullfs":
+            storage_backend = NullFSBasejailStorage.NullFSBasejailStorage
+
+        # if self.config.type == "clonejail":
+        #   pass
+
+        if storage_backend is not None:
+            storage_backend.apply(self.storage, release)
+
+        self.config.fstab.write()
+        self.launch_jail()
+
+        if self.config.vnet:
+            self.start_vimage_network()
+            self.set_routes()
+
+        self.set_nameserver()
+
+        if self.config.jail_zfs is True:
+            ZFSShareStorage.mount_zfs_shares(self.storage)
+
+    def stop(self):
+        self.require_jail_existing()
+        self.require_jail_running()
+        self.destroy_jail()
+        if self.config.vnet:
+            self.stop_vimage_network()
+        self.update_jail_state()
+
+    def create(self, release_name, auto_download=False):
+        self.require_jail_not_existing()
+
+        # check if release exists
+        releases = Releases.Releases(
+            host=self.host, zfs=self.zfs, logger=self.logger)
+        try:
+            release = releases.find_by_name(release_name)
+        except:
+            fetched_release = ", ".join(
+                list(map(lambda x: x.name, releases.local)))
+            msg = f"Can only create from a fetched release ({fetched_release})"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+        self.config.release = release.name
+
+        try:
+            if not isinstance(self.uuid, uuid.UUID):
+                raise
```

Comment:
> Also note that you are checking self.uuid but assigning self.config.uuid, is that intended?

Initially the UUID was noted in the config as well, so that was the right place to store it. Accessing the jail's identifier `uuid` directly made sense to me at the time of writing. In the meanwhile iocage changed the use of Jail identifiers, so I will need to overthink this concept.

@surajrav your requested change will be applied with one of the next commits.